types/react-router/index.d.ts

```diff
@@ -62,34 +63,36 @@ export interface RouteComponentProps<P> {
   staticContext?: any;
 }
 
-export interface RouteProps {
+export type RouteComponentPropsUnion<P> = Partial<RouteComponentProps<any>> & P;
+
+export interface RouteProps<P> {
   location?: H.Location;
-  component?: React.SFC<RouteComponentProps<any> | undefined> | React.ComponentClass<RouteComponentProps<any> | undefined>;
+  component?: React.SFC<RouteComponentPropsUnion<P> | undefined> | React.ComponentClass<RouteComponentPropsUnion<P> | undefined>;
```

Comment:
Are you sure `P` from `RouteProps` can be used for component props constraint?
Also, not sure if updating `RouteComponentProps` to be `Partial` is correct.